// pdfjs-dist/lib/web/stubs-geckoview.js
/**
 * @licstart The following is the entire license notice for the
 * JavaScript code in this page
 *
 * Copyright 2023 Mozilla Foundation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * @licend The above is the entire license notice for the
 * JavaScript code in this page
 */
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Toolbar = exports.SecondaryToolbar = exports.PDFThumbnailViewer = exports.PDFSidebarResizer = exports.PDFSidebar = exports.PDFPresentationMode = exports.PDFOutlineViewer = exports.PDFLayerViewer = exports.PDFFindBar = exports.PDFDocumentProperties = exports.PDFCursorTools = exports.PDFAttachmentViewer = exports.AnnotationEditorParams = void 0;
const AnnotationEditorParams = null;
exports.AnnotationEditorParams = AnnotationEditorParams;
const PDFAttachmentViewer = null;
exports.PDFAttachmentViewer = PDFAttachmentViewer;
const PDFCursorTools = null;
exports.PDFCursorTools = PDFCursorTools;
const PDFDocumentProperties = null;
exports.PDFDocumentProperties = PDFDocumentProperties;
const PDFFindBar = null;
exports.PDFFindBar = PDFFindBar;
const PDFLayerViewer = null;
exports.PDFLayerViewer = PDFLayerViewer;
const PDFOutlineViewer = null;
exports.PDFOutlineViewer = PDFOutlineViewer;
const PDFPresentationMode = null;
exports.PDFPresentationMode = PDFPresentationMode;
const PDFSidebar = null;
exports.PDFSidebar = PDFSidebar;
const PDFSidebarResizer = null;
exports.PDFSidebarResizer = PDFSidebarResizer;
const PDFThumbnailViewer = null;
exports.PDFThumbnailViewer = PDFThumbnailViewer;
const SecondaryToolbar = null;
exports.SecondaryToolbar = SecondaryToolbar;
const Toolbar = null;
exports.Toolbar = Toolbar;
