// compactor/src/util/FileUtils.js


module.exports.PDF_SMALL_COMPRESS_OPTIONS = {
    pageScale: 1.0,
    pageQuality: 0.75,
};

module.exports.PDF_LARGE_COMPRESS_OPTIONS = {
    pageScale: 1.0,
    pageQuality: 0.45,
};

module.exports.JPEG_SMALL_COMPRESS_OPTIONS = {
    pageScale: 1.0,
    pageQuality: 0.1,
};

module.exports.JPEG_LARGE_COMPRESS_OPTIONS = {
    pageScale: 1.0,
    pageQuality: 0.05,
};

module.exports.PNG_SMALL_COMPRESS_OPTIONS = {
    pageScale: 1.0,
    pageQuality: 0.50,
};

module.exports.PNG_LARGE_COMPRESS_OPTIONS = {
    pageScale: 1.0,
    pageQuality: 0.25,
};

module.exports.getFileName = (filename) =>
    filename.slice(0, filename.lastIndexOf(".")).toLowerCase();

module.exports.getImageObject = data => new Promise((resolve, reject) => {
    const image = new Image();
    image.src = data;
    image.onload = () => resolve(image);
    image.onerror = () => reject("Error cannot create the image");
});

module.exports.setCompressedFileFields = (inputFile, base64String) => {
    const base64PreFixLength = getPreFix(inputFile.mimeType).length;

    inputFile.bytes = base64String.substring(base64PreFixLength);
    inputFile.fileSize = getFileSize(base64String);
}

module.exports.createCanvas = (width, height) => {
    const { createCanvas } = require("canvas");
    const canvas = createCanvas(width, height);
    return canvas;
}

module.exports.createCanvasFromViewport = (viewPort) => {
    return createCanvas(viewPort.width || viewPort.viewBox[2], viewPort.height || viewPort.viewBox[3]);
}

module.exports.createCanvasContext = (canvas) => {
    return canvas.getContext('2d');
}

module.exports.getFileExtension = (filename) =>
    filename.slice(filename.lastIndexOf(".") + 1).toLowerCase();
